File: tools/linux/sigtest.c
/* sigtest: Stage-3 regression for AOS Linux signal syscalls:
 *   rt_sigaction (174)          - register a handler for SIGUSR1
 *   tkill (238) / raise         - self-signal delivery
 *   handler frame + rt_sigreturn(173)
 *   sigprocmask (175)           - block SIGUSR2, verify it stays pending
 * Prints SIGTEST OK only if every check passes.
 */
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>

static volatile int sigcnt = 0;
static volatile int got_sig = 0;

static void handler(int s) {
    got_sig = s;
    sigcnt++;
    printf("  in handler sig=%d cnt=%d\n", s, sigcnt);
}

int main(void) {
    printf("sigtest start pid=%d\n", (int)getpid());

    printf("push SIGUSR1 handler...\n");
    if (signal(SIGUSR1, handler) == SIG_ERR) {
        printf("FAIL signal(SIGUSR1)\n");
        printf("SIGTEST FAIL\n");
        return 1;
    }

    printf("raise(SIGUSR1)...\n");
    raise(SIGUSR1);
    printf("after raise sigcnt=%d got=%d\n", sigcnt, got_sig);
    if (sigcnt != 1 || got_sig != SIGUSR1) {
        printf("SIGTEST FAIL (usrcount)\n");
        return 1;
    }

    /* block SIGUSR2, raise it, then unblock and expect the handler to fire
     * exactly once (pending delivered after sigprocmask). */
    printf("block SIGUSR2...\n");
    sigset_t set, old;
    sigemptyset(&set);
    sigaddset(&set, SIGUSR2);
    if (sigprocmask(SIG_BLOCK, &set, &old) != 0) {
        printf("SIGTEST FAIL (sigprocmask block)\n");
        return 1;
    }

    printf("raise(SIGUSR2) while blocked...\n");
    raise(SIGUSR2);                    /* must stay pending */
    if (sigcnt != 1) {
        printf("SIGTEST FAIL (delivered while blocked)\n");
        return 1;
    }

    printf("unblock SIGUSR2...\n");
    if (sigprocmask(SIG_SETMASK, &old, 0) != 0) {
        printf("SIGTEST FAIL (sigprocmask unblock)\n");
        return 1;
    }
    printf("after unblock sigcnt=%d got=%d\n", sigcnt, got_sig);
    if (sigcnt != 2 || got_sig != SIGUSR2) {
        printf("SIGTEST FAIL (pending not delivered)\n");
        return 1;
    }

    printf("SIGTEST OK\n");
    return 0;
}

File: scripts/sigtest.py
#!/usr/bin/env python3
"""sigtest: E2E regression for Stage-3 Linux signal syscalls.

Boots in TEXT mode (console-only) so program stdout reliably returns over
serial, then runs lin/sigtest, which verifies:
  - rt_sigaction (174): register a SIGUSR1 handler
  - tkill (238) / raise: synchronous self-signal delivery + handler frame
  - rt_sigreturn (173): restore context after the handler returns
  - rt_sigprocmask (175): SIGUSR2 blocked (stays pending), then delivered
"""
import sys
import time

from qtest import QTest


def boot_text_entry(q, s):
    """Select the second GRUB entry ("AOS (text)") and wait for a prompt."""
    out = b""
    for _ in range(6):
        time.sleep(1.5)
        q.key("down")
        q.key("ret")
        out += q.serial_drain(s, timeout=3)
        if len(out.strip()) > 20:
            break
    out += q.serial_drain(s, timeout=60)
    if b"AOS>" not in out and b"@aos:" not in out:
        raise AssertionError("shell prompt missing after text-mode boot:\n"
                             + out[-400:].decode(errors="replace"))
    return out


def main():
    with QTest("sigtest", serial_mode="socket", autoboot_grub=False) as q:
        s = q.serial_socket()
        boot_text_entry(q, s)

        s.sendall(b"lin/sigtest\r")
        resp = b""
        deadline = time.time() + 60
        while time.time() < deadline:
            resp += q.serial_drain(s, timeout=2, needle=b"SIGTEST OK")
            if b"SIGTEST OK" in resp or b"SIGTEST FAIL" in resp or \
               b"KERNEL PANIC" in resp:
                break
        resp += q.serial_drain(s, timeout=3)
        if b"KERNEL PANIC" in resp:
            raise AssertionError("kernel panic in lin/sigtest:\n"
                                 + resp[-500:].decode(errors="replace"))
        if b"SIGTEST FAIL" in resp or b"FAIL " in resp:
            raise AssertionError("lin/sigtest reported failures:\n"
                                 + resp[-900:].decode(errors="replace"))
        if b"SIGTEST OK" not in resp:
            raise AssertionError("lin/sigtest missing SIGTEST OK; out:\n"
                                 + resp[-900:].decode(errors="replace"))
        print("PASS: lin/sigtest (rt_sigaction + tkill + sigprocmask)")
        return 0


if __name__ == "__main__":
    sys.exit(main())
